hnn_core/optimization.py

```diff
@@ -480,12 +484,22 @@ def optimize_evoked(net, tstop, n_trials, target_dpl, initial_dpl, maxiter=50,
 
     net = net.copy()
 
-    drive_names = [key for key in net.external_drives.keys()
-                   if net.external_drives[key]['type'] == 'evoked']
-    if len(drive_names) == 0:
+    evoked_drive_names = [key for key in net.external_drives.keys()
+                          if net.external_drives[key]['type'] == 'evoked']
+
+    if len(evoked_drive_names) == 0:
         raise ValueError('The current Network instance lacks any evoked '
                          'drives. Consider adding drives using '
                          'net.add_evoked_drive')
+    elif which_drives == 'all':
+        drive_names = evoked_drive_names
+    else:
+        drive_names = [mydrive for mydrive in np.unique(which_drives)
+                       if mydrive in evoked_drive_names]
+    if len(drive_names) == 0:
+        raise ValueError('The drives selected to be optimized are not evoked '
+                         'drives. Optimization works only evoked drives.')
```

Comment:
This looks good! For brevity's sake, we could probably combine the initial check for `len(evoked_drive_names == 0` into the final check for `len(drive_names) == 0` and raise a more general error about no available drives being selected, but that is completely a stylistic suggestion. I'll let @mohdsherif and @jasmainak have the final say here.

Reply:
I actually like the separate messages. It's a bit more informative for the user :)